sklearn/cross_validation.py

```diff
@@ -1264,58 +1253,74 @@ def cross_val_predict(estimator, X, y=None, cv=None, n_jobs=1,
                         pre_dispatch=pre_dispatch)
     preds_blocks = parallel(delayed(_fit_and_predict)(clone(estimator), X, y,
                                                       train, test, verbose,
-                                                      fit_params)
+                                                      fit_params, predict_function)
                             for train, test in cv)
 
-    preds = [p for p, _ in preds_blocks]
+    # obtain original indexes from parallelized predictions
     locs = np.concatenate([loc for _, loc in preds_blocks])
     if not _check_is_partition(locs, _num_samples(X)):
         raise ValueError('cross_val_predict only works for partitions')
     inv_locs = np.empty(len(locs), dtype=int)
     inv_locs[locs] = np.arange(len(locs))
 
-    # Check for sparse predictions
-    if sp.issparse(preds[0]):
-        preds = sp.vstack(preds, format=preds[0].format)
+    # create predictions with correct indexes
+    if isinstance(predict_function, str):
+        concat_pred = _concat_arrays([p for p, _ in preds_blocks])
+        preds = concat_pred[inv_locs]
     else:
-        preds = np.concatenate(preds)
-    return preds[inv_locs]
+        func_preds = list()
+        for func_pred in zip(*(p_tuple for p_tuple, _ in preds_blocks)):
+            concat_pred = _concat_arrays(func_pred)
+            func_preds.append(concat_pred[inv_locs])
+        preds = tuple(func_preds)
+    return preds
 
 
-def _fit_and_predict(estimator, X, y, train, test, verbose, fit_params):
-    """Fit estimator and predict values for a given dataset split.
+def _concat_arrays(arrays):
```

Comment:
This was factored out of the existing implementation https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/cross_validation.py#L1277 so it could be reused in my implementation.

A quick search of `vstack` in the existing master branch did not turn up any related functions in utils. It would make sense to move this there.